src/parser/mod.rs

```diff
@@ -3,34 +3,34 @@ macro_rules! get_expanders {
     ($vars:expr, $dir_stack:expr) => {
         ExpanderFunctions {
             tilde: &|tilde: &str| $vars.tilde_expansion(tilde, $dir_stack),
-            array: &|array: &str, index: Index| {
+            array: &|array: &str, selection : Select| {
                 use std::iter::FromIterator;
                 use $crate::types::*;
 
                 match $vars.get_array(array) {
-                    Some(array) => match index {
-                        Index::None   => None,
-                        Index::All    => Some(array.clone()),
-                        Index::ID(id) => array.get(id).map(
-                            |x| Array::from_iter(Some(x.to_owned()))
-                        ),
-                        Index::FromEnd(idx) => {
-                            if array.len() < idx {
-                                None
-                            } else {
-                                array.get(array.len() - idx)
-                                     .map(|x| Array::from_iter(Some(x.to_owned())))
-                            }
+                    Some(array) => match selection {
+                        Select::None  => None,
+                        Select::All   => Some(array.clone()),
+                        Select::Index(id) => {
+                            id.resolve(array.len())
+                              .and_then(|n| array.get(n))
+                              .map(|x| Array::from_iter(Some(x.to_owned())))
```

Comment:
There are a bunch of places where `ion` uses strange syntax in order to create a singleton `SmallVec`. Maybe its time to add in a `smallvec![...]` macro?